modules/core/04-channel/keeper/packet.go

```diff
@@ -133,10 +140,16 @@ func (k Keeper) RecvPacket(
 		return errorsmod.Wrap(types.ErrChannelNotFound, packet.GetDestChannel())
 	}
 
-	if channel.State != types.OPEN {
+	counterpartyLastPacketSent, _ := k.GetCounterpartyLastPacketSequence(ctx, packet.GetDestPort(), packet.GetDestChannel())
+
+	if channel.State != types.OPEN && !(channel.FlushStatus == types.FLUSHING && packet.GetSequence() <= counterpartyLastPacketSent) {
 		return errorsmod.Wrapf(
 			types.ErrInvalidChannelState,
-			"channel state is not OPEN (got %s)", channel.State.String(),
+			"packets cannot be received: either channel with non OPEN state (%s) & not in upgrade, or channel is in upgrade but flush status (%s) is not FLUSHING & packet sequence %d must be less than or equal to counterparty last packet sent sequence %d",
+			channel.State,
+			channel.FlushStatus,
+			packet.GetSequence(),
+			counterpartyLastPacketSent,
```

Comment:
I definitely found this section pretty hard to parse and reason about. Looks like there was already some discussion in another thread about it. 

I find it a bit complex to mix a channel state assertion with the `&&` expression which opens the door for weird channel states to slip through. So here, we are only going to fail the recv packet tx if the channel state is not OPEN AND the channel state is not flushing and the sequencing is all good. 

So if we are in UPGRADETRY or UPGRADEACK, or any other state for that matter then the first part of the expression will evaluate as false.

What do you think about explicitly checking the expected channel states? I would be more confident in the code and see less room for unexpected states to slip through the cracks with something like:

```go
if !collections.Contains(channel.State, []types.State{types.OPEN, types.TRYUPGRADE, types.ACKUPGRADE}) {
	return errorsmod.Wrapf(types.ErrInvalidChannelState, "channel state is not OPEN (got %s)", channel.State.String())
}

if counterpartyLastSequenceSend, found := k.GetCounterpartyLastPacketSequence(ctx, packet.GetDestPort(), packet.GetDestChannel()); found {
	if channel.FlushStatus != types.FLUSHING && packet.GetSequence() <= counterpartyLastSequenceSend {
		return errorsmod.Wrap(types.ErrInvalidChannelState, "something meaningful")
	}
}
```

I think we can also make use of the `found` bool with last sequence send instead of discarding it

Reply:
Slightly related https://github.com/cosmos/ibc-go/issues/4001 - I noticed this field was never being set in store within the upgrade handshake

Reply:
I'm a big fan of this change actually, it makes things far more readable. =

Reply:
note the last conditional should be `||`, righto?

Reply:
I think it might also make sense to add a comment that the second check for the last seq sent is only meaningful for `UPGRADETRY/ACK`.

Reply:
Currently the tests pass with both `||` and `&&` so I think we don't have enough test coverage 😅

Reply:
But yes I believe `||` is correct here! If either of those conditions are false we should error.

Reply:
also, for anyone else reviewing, the original condition was completely flipped with Damians suggestion, hence why we can break the giant if in two smaller ones. `&&` cannot be broken directly like this but since it was flipped to an `||` it can, this kept me up a bit.